src/ast/irbuilderbpf.cpp

```diff
@@ -819,6 +819,146 @@ Value *IRBuilderBPF::CreateStrncmp(Value *str1,
 
   return result;
 }
+Value *IRBuilderBPF::CreateStrstr(Value *val1,
+                                  uint64_t str1_size,
+                                  Value *val2,
+                                  uint64_t str2_size,
+                                  bool inverse)
+{
+  /*
+  // This function compares whether the string val1 contains the string val2.
+  // It returns true if val2 is contained by val1, false if not contained.
+  // strstr(String val1, String val2, int str1_size, int str2_size)
+     {
+        for (size_t j = 0; (str1_size >= str2_size) && (j <= str1_size -
+  str2_size); j++)
+        {
+          for (size_t i = 0; i < str2_size; i++)
+          {
+            if (val2[i] == NULL)
+            {
+              return true;
+            }
+            if (val1[i + j] != val2[i])
+            {
+              break;
+            }
+          }
+          if (val1[j] == NULL) {
+            return false;
+          }
+        }
+        return false;
+     }
+  */
+  // Check if the compared strings are literals.
+  // If so, we can avoid storing the literal in memory.
+  std::optional<std::string> literal1;
+  if (auto constString1 = dyn_cast<ConstantDataArray>(val1))
+    literal1 = constString1->getAsString();
+  else if (isa<ConstantAggregateZero>(val1))
+    literal1 = "";
+  else
+    literal1 = std::nullopt;
+
+  std::optional<std::string> literal2;
+  if (auto constString2 = dyn_cast<ConstantDataArray>(val2))
+    literal2 = constString2->getAsString();
+  else if (isa<ConstantAggregateZero>(val2))
+    literal2 = "";
+  else
+    literal2 = std::nullopt;
```

Comment:
We should handle both strings being literal and not even emit the branches below.

Reply:
sorry, I don't understand what you mean. here I just mimic the implementation of strncmp.

Reply:
If both strings are literal, you can directly determine the result here (e.g. by using `std::string::find`) and return it.

Reply:
Yes, it could use the same optimization.